src/CompletionProvider.php

```diff
@@ -372,38 +393,37 @@ private function suggestVariablesAtNode(Node $node, string $namePrefix = ''): ar
      *
      * @param Node $node
      * @param string $namePrefix Prefix to filter
-     * @return Node\Expr\Variable[]
+     * @return Node\Expression\Variable[]
      */
     private function findVariableDefinitionsInNode(Node $node, string $namePrefix = ''): array
     {
         $vars = [];
         // If the child node is a variable assignment, save it
-        $parent = $node->getAttribute('parentNode');
-        if (
-            $node instanceof Node\Expr\Variable
-            && ($parent instanceof Node\Expr\Assign || $parent instanceof Node\Expr\AssignOp)
-            && is_string($node->name) // Variable variables are of no use
-            && substr($node->name, 0, strlen($namePrefix)) === $namePrefix
-        ) {
-            $vars[] = $node;
-        }
-        // Iterate over subnodes
-        foreach ($node->getSubNodeNames() as $attr) {
-            if (!isset($node->$attr)) {
-                continue;
-            }
-            $children = is_array($node->$attr) ? $node->$attr : [$node->$attr];
-            foreach ($children as $child) {
-                // Dont try to traverse scalars
-                // Dont traverse functions, the contained variables are in a different scope
-                if (!($child instanceof Node) || $child instanceof Node\FunctionLike) {
-                    continue;
-                }
-                foreach ($this->findVariableDefinitionsInNode($child, $namePrefix) as $var) {
-                    $vars[] = $var;
+
+        $isAssignmentToVariable = function ($node) use ($namePrefix) {
+            return $node instanceof Node\Expression\AssignmentExpression
+                && $node->leftOperand instanceof Node\Expression\Variable
+                && (empty($namePrefix) || strpos($node->leftOperand->getName(), $namePrefix) !== false);
+        };
+        $isNotFunctionLike = function ($node) {
```

Comment:
With pure helper function, I meant a stand-alone function declaration instead of an anonymous function. Ideally `isAssignmentToVariable` is also made pure and changed.